### PR TITLE
Added jdbc autoconfiguration

### LINKED ISSUE #337: OpenTracing JDBC Auto Configuration
## Detailed Description

Define a module providing a Spring Boot Auto Configuration for `opentracing-jdbc`.

## Context

- Integration into Spring Boot ecosystem
- Coupling to Spring is encapsulated in separate module/artifact
- Zalando Conventions over Configuration, i.e. ready to use
- Customizability via Spring Boot configuration and beans

### Module
- `opentracing-jdbc` (parent) :heavy_check_mark: 
  - `opentracing-jdbc` :heavy_check_mark: 
  - `opentracing-jdbc-autoconfigure` :new: 

### Auto Configuration
- `OpenTracingDataSourceAutoConfiguration`
- Apply to all data sources by default?
- Wrap data sources in `@Primary` beans?!
- Find registered `SpanDecorator` implementations (inject `@Order`ed list)

### PR DESCRIPTION
**Pull request by whiskeysierra**
## Motivation and Context

Fixes #337

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/CustomDataSourceTracerTest.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+package org.zalando.opentracing.jdbc.autoconfigure;
+
+import io.opentracing.Tracer;
+import io.opentracing.mock.MockSpan;
+import io.opentracing.mock.MockTracer;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.platform.commons.util.CollectionUtils;
```
Comment (whiskeysierra) [inline opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/CustomDataSourceTracerTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit.platform.commons.util.CollectionUtils'](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701844)

File: opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import javax.sql.DataSource;
```
Comment (whiskeysierra) [inline opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfigurationTest.java]: Codacy Issue found: [Avoid unused imports such as 'javax.sql.DataSource'](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701844)

File: opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/CustomDataSourceTracerTest.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+import org.zalando.opentracing.jdbc.DataSourceTracer;
+import org.zalando.opentracing.jdbc.span.ComponentSpanDecorator;
+
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
```
Comment (whiskeysierra) [inline opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/CustomDataSourceTracerTest.java]: Codacy Issue found: [Avoid unused imports such as 'java.util.List'](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701844)

File: opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/CustomDataSourceTracerTest.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.Map;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.aMapWithSize;
+import static org.hamcrest.Matchers.hasEntry;
+import static org.hamcrest.Matchers.hasSize;
```
Comment (whiskeysierra) [inline opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/CustomDataSourceTracerTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.hamcrest.Matchers.hasSize'](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701844)

File: opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfigurationTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    @Autowired
+    private HikariDataSource dataSource;
+
+    @Test
+    void proxiesDataSource() {
+        assertTrue(isAopProxy(dataSource));
+    }
+
+    @Test
+    void supportsDirectAccess() throws SQLException {
```
Comment (whiskeysierra) [inline opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfigurationTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701844)

File: opentracing-jdbc/opentracing-jdbc-autoconfigure/src/main/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            return new DataSourceTracer(tracer)
+                    .withOperationName(operationName)
+                    .withLifecycle(lifecycle)
+                    .withActivation(activation)
+                    .withSpanDecorators(composite(decorators));
+        }
+
+    }
+
+    // TODO how can we give the user more control to decide which data sources
```
Comment (lukasniemeier-zalando) [inline opentracing-jdbc/opentracing-jdbc-autoconfigure/src/main/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfiguration.java]: What about allowing to deactivate just this post processed via a conditional property? Then clients have the power to turn off annotating all data sources and can decide themselves.

File: opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfigurationTest.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    private HikariDataSource dataSource;
+
+    @Test
+    void proxiesDataSource() {
+        assertTrue(isAopProxy(dataSource));
+    }
+
+    @Test
+    void supportsDirectAccess() {
+        dataSource.isRunning();
```
Comment (lukasniemeier-zalando) [inline opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfigurationTest.java]: todo: use (un)proxied method with parameter(s)

File: opentracing-jdbc/opentracing-jdbc-autoconfigure/src/main/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfiguration.java (hunk 6 above)
Comment (whiskeysierra) [reply opentracing-jdbc/opentracing-jdbc-autoconfigure/src/main/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfiguration.java]: Done

File: opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfigurationTest.java (hunk 7 above)
Comment (whiskeysierra) [reply opentracing-jdbc/opentracing-jdbc-autoconfigure/src/test/java/org/zalando/opentracing/jdbc/autoconfigure/OpenTracingJdbcAutoConfigurationTest.java]: Done

**Comment by lukasniemeier-zalando** [comment]
:+1:

**Pull request merged**